### PR TITLE
feat(mdns): Add support for IPv4 reverse query

### PR DESCRIPTION
**Pull request by david-cermak**
~~WIP: Need to create delegated name with the reverse query:~~

Reverse queries need to be enabled in menuconfig and registered when IP changes, typically within an `IP_EVENT` handler:

For user defined interfaces, we just add the registration event to others:
```cpp
ESP_ERROR_CHECK(mdns_netif_action(esp_netif, MDNS_EVENT_ENABLE_IP4 | MDNS_EVENT_IP4_REVERSE_LOOKUP));
```

For predefined interfaces, we have to enable it explicitly, after the interface is registered in `mdns` internals

Partially addresses https://github.com/espressif/esp-protocols/issues/101

### CODE REVIEW

File: components/mdns/tests/host_test/components/esp_event/CMakeLists.txt [hunk 1]
```diff
@@ -1,3 +1,3 @@
 idf_component_register(SRCS esp_event_mock.c
                        INCLUDE_DIRS include
-                       REQUIRES esp_system_protocols_linux)
+                       REQUIRES)
```
Comment (euripedesrocha) [inline components/mdns/tests/host_test/components/esp_event/CMakeLists.txt:3]:
```suggestion
                       )
```

File: components/mdns/tests/host_test/components/esp_event/include/esp_event.h [hunk 2]
```diff
@@ -8,16 +8,12 @@
 #include "stdbool.h"
 #include "esp_err.h"
 #include "esp_event_base.h"
-#include "bsd_strings.h"
+//#include "bsd_strings.h"
```
Comment (euripedesrocha) [inline components/mdns/tests/host_test/components/esp_event/include/esp_event.h:11]: Should this be deleted?

File: components/mdns/tests/host_test/components/esp_event/include/esp_event.h [hunk 3]
```diff
@@ -8,16 +8,12 @@
...
 
-#define ESP_EVENT_DECLARE_BASE(x)
-#define ESP_EVENT_ANY_ID       (-1)
 
-typedef void *esp_event_base_t;
-typedef void *system_event_t;
 
-const char *WIFI_EVENT;
-const char *IP_EVENT;
+//const char *WIFI_EVENT;
```
Comment (euripedesrocha) [inline components/mdns/tests/host_test/components/esp_event/include/esp_event.h:15]: Should this be deleted?

File: components/mdns/tests/host_test/components/esp_timer_linux/timer_task.hpp [hunk 4]
```diff
@@ -16,7 +16,7 @@ typedef void (*cb_t)(void *arg);
 
 class TimerTaskMock {
 public:
-    TimerTaskMock(cb_t cb): cb(cb), t(run_static, this), active(false), ms(INT32_MAX) {}
+    TimerTaskMock(cb_t cb): cb(cb), active(false), ms(INT32_MAX) {}
```
Comment (euripedesrocha) [inline components/mdns/tests/host_test/components/esp_timer_linux/timer_task.hpp:19]: This class could make use of std::jthread.

File: components/mdns/tests/host_test/components/freertos/CMakeLists.txt [hunk 5]
```diff
@@ -1,6 +1,6 @@
 idf_component_register(SRCS freertos_linux.c queue_unique_ptr.cpp
                        INCLUDE_DIRS include
-                       REQUIRES esp_system_protocols_linux)
+                       REQUIRES)
```
Comment (euripedesrocha) [inline components/mdns/tests/host_test/components/freertos/CMakeLists.txt:3]:
```suggestion
                       )
```

File: components/mdns/tests/host_test/components/freertos/freertos_linux.c [hunk 6]
```diff
@@ -158,15 +158,15 @@ void xTaskCreate(TaskFunction_t pvTaskCode, const char *const pcName, const uint
...
-uint32_t esp_get_free_heap_size(void)
-{
-    return 0;
-}
-
-uint32_t esp_random(void)
-{
-    return rand();
-}
+//uint32_t esp_get_free_heap_size(void)
```
Comment (euripedesrocha) [inline components/mdns/tests/host_test/components/freertos/freertos_linux.c:161]: Delete?

File: components/mdns/tests/host_test/sdkconfig.defaults [hunk 7]
```diff
@@ -0,0 +1,9 @@
+CONFIG_IDF_TARGET="linux"
+CONFIG_MDNS_NETWORKING_SOCKET=y
+CONFIG_MDNS_SKIP_SUPPRESSING_OWN_QUERIES=y
+CONFIG_TEST_NETIF_NAME="eth0"
+CONFIG_TEST_HOSTNAME="myesp"
+# CONFIG_MDNS_PREDEF_NETIF_STA is not set
```
Comment (euripedesrocha) [inline components/mdns/tests/host_test/sdkconfig.defaults:6]: Delete?

File: .github/workflows/host-test.yml [hunk 8]
```diff
@@ -3,8 +3,8 @@ name: Host test
 on: [push, pull_request]
 
 jobs:
-  host_test:
-    name: Build and Test on Host
+  host_test_esp_modem:
+    name: esp-modem Build and Test on Host
     runs-on: ubuntu-20.04
     container: espressif/idf:release-v4.3
```
[the remark this answers is not in the document]
Comment (david-cermak) [reply .github/workflows/host-test.yml:9]: `esp_modem` linux/mock layers are quite old and work only with this IDF version. I'm planning to update it per latest IDF (which would definitely simplify it), but haven't gotten to that yet.

File: components/mdns/mdns.c [hunk 9]
```diff
@@ -17,10 +17,15 @@
 #include "mdns_networking.h"
 #include "esp_log.h"
 #include "esp_random.h"
-#if CONFIG_ETH_ENABLED
+#if __has_include("bsd/string.h")
```
[the remark this answers is not in the document]
Comment (david-cermak) [reply components/mdns/mdns.c]: Yes, in when building with standard *real-target* IDF, we can just include `string.h`. On host, we'd have to install `libbsd` and include their `string.h` this way.

But I agree that it doesn't look nice in the production code. Have removed it and solving it differently in the port layers.

File: components/mdns/mdns.c [hunk 10]
```diff
@@ -639,6 +647,53 @@ static inline int append_one_txt_record_entry(uint8_t *packet, uint16_t *index,
...
+    char *start;
+    do  {
+        start = end;
+        end = memchr(start, '.', len);
+        end = end ? end : host + len;
+        int part_len = end - start;
+        if ((*index + part_len + 1) >= MDNS_MAX_PACKET_SIZE) {
+            return 0;
+        }
+        _mdns_append_u8(packet, index, part_len);
```
[the remark this answers is not in the document]
Comment (david-cermak) [reply components/mdns/mdns.c]: Thanks, fixed!

File: components/mdns/mdns.c [hunk 11]
```diff
@@ -639,6 +647,53 @@ static inline int append_one_txt_record_entry(uint8_t *packet, uint16_t *index,
...
+        memcpy(packet + *index, start, part_len);
+        *index += part_len;
+        end++;
+    } while (end < name + len);
+
+    if (last) {
+        return *index;
+    }
+    if (!append_fqdn_dots(packet, index, "arpa", true)) {
```
[the remark this answers is not in the document]
Comment (david-cermak) [reply components/mdns/mdns.c:686]: Let's talk about this more in a separate discussion.

File: components/mdns/mdns.c [hunk 12]
```diff
@@ -639,6 +647,53 @@ static inline int append_one_txt_record_entry(uint8_t *packet, uint16_t *index,
...
+ * where we gain very little (or compression even gets counter-productive mainly for IPv6 addresses)
+ *
+ * @param  packet       MDNS packet
+ * @param  index        offset in the packet
+ * @param  name         name representing FQDN in '.' separated parts
+ * @param  last         true if appending the last part (domain, typically "arpa")
+ *
+ * @return length of added data: 0 on error or length on success
+ */
+static uint16_t append_fqdn_dots(uint8_t *packet, uint16_t *index, const char *name, bool last)
```
[the remark this answers is not in the document]
Comment (david-cermak) [reply components/mdns/mdns.c:661]: Good point and WIP already.

Added a new PR with unit test support and @antmak has put up [this PR](https://github.com/espressif/esp-protocols/pull/193) to decouple some mdns layers so they could be separately unit-testable.

But, that's not going to happen in this PR

File: components/mdns/mdns.c [hunk 13]
```diff
@@ -1200,6 +1264,37 @@ static uint8_t _mdns_append_host_answer(uint8_t *packet, uint16_t *index, mdns_h
...
+
+    if (!append_fqdn_dots(packet, index, name, false)) {
+        return 0;
+    }
+
+    if (!_mdns_append_type(packet, index, MDNS_ANSWER_PTR, false, 10 /* TTL set to 10s*/ )) {
+        return 0;
+    }
+
+    uint16_t data_len_location = *index - 2;
```
[the remark this answers is not in the document]
Comment (david-cermak) [reply components/mdns/mdns.c]: Thanks, added a comment!

File: components/mdns/mdns.c [hunk 14]
```diff
@@ -1200,6 +1264,37 @@ static uint8_t _mdns_append_host_answer(uint8_t *packet, uint16_t *index, mdns_h
...
+    const char *str[2] = { _mdns_self_host.hostname, MDNS_DEFAULT_DOMAIN };
+
+    int part_length = _mdns_append_fqdn(packet, index, str, 2, MDNS_MAX_PACKET_SIZE);
```
[the remark this answers is not in the document]
Comment (david-cermak) [reply components/mdns/mdns.c:1288]: From the description of `append_fqdn_dots()`:

```cpp
appends FQDN to a packet from hostname separated by dots.
```

So, yes, it adds one FQDN. The function `_mdns_append_reverse_ptr_record()`  however creates an entire record, which consists of
* One FQDN (separated by dots ~ the address we're trying to resolve)
* type/class etc (per DNS spec)
* One FQDN (from substrings ~ the name that owns the queried address)

File: components/mdns/tests/host_test/components/esp_event/include/esp_event.h (hunk 3 above)
Comment (david-cermak) [reply components/mdns/tests/host_test/components/esp_event/include/esp_event.h]: Thanks, fixed

File: components/mdns/tests/host_test/sdkconfig.defaults (hunk 7 above)
Comment (david-cermak) [reply components/mdns/tests/host_test/sdkconfig.defaults:6]: Well, this is a bit confusing, but cannot delete this. According to the `sdkconfig` spec this means that the `CONFIG_MDNS_PREDEF_NETIF_STA` should be disabled.

File: components/mdns/tests/host_test/make_eth2_netif [hunk 15]
```diff
@@ -0,0 +1,5 @@
+sudo ip link add eth2 type dummy
```
Comment (kostaond) [inline components/mdns/tests/host_test/make_eth2_netif:1]: Is this to be bash script? Or is somewhere defined how to use this file?

File: components/mdns/tests/host_test/main/main.c [hunk 16]
```diff
@@ -35,14 +35,16 @@ int main(int argc, char *argv[])
 {
 
     setvbuf(stdout, NULL, _IONBF, 0);
-    const esp_netif_inherent_config_t base_cg = { .if_key = "WIFI_STA_DEF", .if_desc = "eth2" };
+    const esp_netif_inherent_config_t base_cg = { .if_key = "WIFI_STA_DEF", .if_desc = CONFIG_TEST_NETIF_NAME };
     esp_netif_config_t cfg = { .base = &base_cg  };
     esp_netif_t *sta = esp_netif_new(&cfg);
+    ESP_ERROR_CHECK(mdns_init());
+    mdns_hostname_set(CONFIG_TEST_HOSTNAME);
```
Comment (gabsuren) [inline components/mdns/tests/host_test/main/main.c:42]: ESP_ERROR_CHECK ?

File: components/mdns/mdns.c [hunk 17]
```diff
@@ -639,6 +645,55 @@ static inline int append_one_txt_record_entry(uint8_t *packet, uint16_t *index,
...
+ * @param  last         true if appending the last part (domain, typically "arpa")
+ *
+ * @return length of added data: 0 on error or length on success
+ */
+static uint16_t append_fqdn_dots(uint8_t *packet, uint16_t *index, const char *name, bool last)
+{
+    int len = strlen(name);
+    char *host = (char *)name;
+    char *end = host;
+    char *start;
```
Comment (david-cermak) [inline components/mdns/mdns.c:666]:
```suggestion
    char *start = host;
```

File: components/mdns/tests/host_test/sdkconfig.defaults (hunk 7 above)
Comment (david-cermak) [reply components/mdns/tests/host_test/sdkconfig.defaults:6]:
```suggestion
CONFIG_MDNS_PREDEF_NETIF_STA=n
```

File: components/mdns/tests/host_test/make_eth2_netif (hunk 15 above)
Comment (david-cermak) [reply components/mdns/tests/host_test/make_eth2_netif:1]: Add shebang line (and add sh extension)

File: components/mdns/tests/host_test/main/main.c (hunk 16 above)
Comment (david-cermak) [reply components/mdns/tests/host_test/main/main.c:42]: will add

File: components/mdns/tests/host_test/make_eth2_netif (hunk 15 above)
Comment (david-cermak) [reply components/mdns/tests/host_test/make_eth2_netif]: Removed the script, as these commands are listed in the `README` and the script is not used in the CI

File: components/mdns/mdns.c (hunk 11 above)
[the remark this answers is not in the document]
Comment (david-cermak) [reply components/mdns/mdns.c]: Split into two statics to get rid of the regression.

File: components/mdns/tests/host_test/components/esp_timer_linux/timer_task.hpp (hunk 4 above)
Comment (david-cermak) [reply components/mdns/tests/host_test/components/esp_timer_linux/timer_task.hpp:19]: I would keep this goal for future, as using `jthread` needs C++20 and I had difficulties enabling in simply in CMake and supporting it in CI runner.

**Comment by david-cermak** [comment]
@euripedesrocha Could you please check the unresolved discussions?

@gabsuren @kostaond Thanks for taking a look, could you please look again and complete your review?

**Comment by gabsuren** [comment]
@david-cermak would be nice to update example tests although, but it can be done separately. LGTM now

**Review by gabsuren — APPROVED** [review]

**Review by euripedesrocha — APPROVED** [review]

**Comment by david-cermak** [comment]
Thanks for the review guys!

> would be nice to update example tests

Good point, but let me merge this first and I'll add the reverse lookup to our example tests once we support IPv6 query.

**Pull request merged**